servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/KeepAliveManager.java

```diff
@@ -0,0 +1,320 @@
+            int idleInSeconds = (int) keepAlivePolicy.idleDuration().getSeconds();
+            idlenessDetector.configure(channel, idleInSeconds, this::channelIdle);
+        } else {
+            disallowKeepAliveWithoutActiveStreams = false;
+            pingAckTimeoutMillis = DEFAULT_ACK_TIMEOUT.toMillis();
+        }
+    }
+
+    void pingReceived(final Http2PingFrame pingFrame) {
+        assert channel.eventLoop().inEventLoop();
+
+        if (pingFrame.ack()) {
+            long pingAckContent = pingFrame.content();
+            if (pingAckContent == GRACEFUL_CLOSE_PING_CONTENT) {
+                LOGGER.debug("channel={}, graceful close ping ack received.", channel);
+                cancelIfStateIsAFuture(gracefulCloseState);
+                gracefulCloseWriteSecondGoAway();
+            } else if (pingAckContent == KEEP_ALIVE_PING_CONTENT) {
+                cancelIfStateIsAFuture(keepAliveState);
+                keepAliveState = null;
+            }
+        } else {
+            // Send an ack for the received ping
+            channel.writeAndFlush(new DefaultHttp2PingFrame(pingFrame.content(), true));
+        }
+    }
+
+    void trackActiveStream(final Channel streamChannel) {
+        activeChildChannelsUpdater.incrementAndGet(this);
+        streamChannel.closeFuture().addListener(f -> {
+            activeChildChannelsUpdater.decrementAndGet(this);
+            if (activeChildChannels == 0 && gracefulCloseState == GRACEFUL_CLOSE_SECOND_GO_AWAY_SENT) {
+                close0();
+            }
+        });
+    }
+
+    void channelClosed() {
+        assert channel.eventLoop().inEventLoop();
+
+        cancelIfStateIsAFuture(gracefulCloseState);
+        cancelIfStateIsAFuture(keepAliveState);
+        gracefulCloseState = CLOSED;
+        keepAliveState = CLOSED;
+    }
+
+    void initiateGracefulClose(final Runnable whenInitiated) {
+        EventLoop eventLoop = channel.eventLoop();
+        if (eventLoop.inEventLoop()) {
+            doCloseAsyncGracefully0(whenInitiated);
+        } else {
+            try {
+                eventLoop.execute(() -> doCloseAsyncGracefully0(whenInitiated));
+            } catch (Throwable cause) {
+                close0();
+                LOGGER.warn("channel={} EventLoop rejected a task for graceful shutdown, force closing connection",
+                        channel, cause);
+            }
+        }
+    }
+
+    void channelIdle() {
+        assert channel.eventLoop().inEventLoop();
+
+        if (keepAliveState == CLOSED) {
+            return;
+        }
+        if (activeChildChannels == 0 && disallowKeepAliveWithoutActiveStreams) {
+            return;
+        }
+        // idleness detected for the first time, send a ping to detect closure, if any.
+        keepAliveState = KEEP_ALIVE_ACK_PENDING;
+        channel.writeAndFlush(new DefaultHttp2PingFrame(KEEP_ALIVE_PING_CONTENT, false))
+                .addListener(future -> {
```

Comment:
consider caching the `ChannelFutureListener` to reduce GC pressure if you need to write the next ping again